lib/promscrape/discovery/kubernetes/api_watcher.go

```diff
@@ -570,6 +579,9 @@ func (uw *urlWatcher) unsubscribeAPIWatcherLocked(aw *apiWatcher) {
 		delete(uw.aws, aw)
 		metrics.GetOrCreateCounter(fmt.Sprintf(`vm_promscrape_discovery_kubernetes_subscribers{role=%q,status="working"}`, uw.role)).Dec()
 	}
+	if (len(uw.aws) + len(uw.awsPending)) == 0 {
```

Comment:
Currently, another bug is possible. If api watcher resubscribes for cancelled urlWatcher, it'll receive an error.

It looks, that ReferenceCount for `urlWatcher` object should work better:

Each `subscribeAPIWatcherLocked` call increases RC.
Each `unsubscribeAPIWatcherLocked` call decreases RC.

At RC == 0, urlWatcher must stopped and deleted.

```go
func (gw *groupWatcher) unsubscribeAPIWatcher(aw *apiWatcher) {
	gw.mu.Lock()
	defer gw.mu.Unlock()
        var deleteKeys []string
	for key, uw := range gw.m {
		uw.unsubscribeAPIWatcherLocked(aw)
                if uw.refCount == 0 {
                    uw.close() // cancel and wait for finish?
                    deleteKeys = append(deleteKeys, key)
                }
	}
        for _, dk := range deleteKeys{
             delete(gw.m,dk)
        }
}
```

WDYT?

Reply:
Implemented this